// hyper/src/app/pod.js
/*/////////////////////////////////////////////////////////////////////////////

  Author  : @qodeninja

/////////////////////////////////////////////////////////////////////////////*/

"use strict";


/*/////////////////////////////////////////////////////////////////////////////
// λ - Includes
/////////////////////////////////////////////////////////////////////////////*/  
  
  const util    = require('util');
  const EE      = require('events');
  const http    = require('http');
  const path    = require('path');


	
/*/////////////////////////////////////////////////////////////////////////////
// λ - Server Prototype
/////////////////////////////////////////////////////////////////////////////*/  

  class HyperPod extends EE{

    constructor( options = {} ){
      super();
      this.config = options;
      this.server = null;
      setTimeout( ()=>{
        this.init();
        this.start();
      },1000);
    }

    init(){
      console.log('stub::init');
      this.emit('init',{today:1});
    }

    conf( options ){
      console.log('stub::config');
      this.config = {...this.config, ...options};
      this.emit('config');
    }

    mount( name, app ){
      console.log('stub::mount');
    }

    start(){
      console.log('stub::start');
      this.emit('start');
    }


  }


/*/////////////////////////////////////////////////////////////////////////////
// Exports
/////////////////////////////////////////////////////////////////////////////*/


  module.exports = HyperPod;


/*/////////////////////////////////////////////////////////////////////////////
// Standalone
/////////////////////////////////////////////////////////////////////////////*/

  if( require.main === module ){
    var args = process.argv.slice(2);
    args.forEach(function (val, index, array) {
      console.log(index + ': ' + val);
    });
  }
